src/ic-cdk/src/api/call.rs

```diff
@@ -68,11 +73,60 @@ impl Future for CallFuture {
 
     fn poll(self: Pin<&mut Self>, context: &mut Context<'_>) -> Poll<Self::Output> {
         let self_ref = Pin::into_inner(self);
+        let state_ptr = Arc::into_raw(self_ref.state.clone());
         let mut state = self_ref.state.write().unwrap();
 
         if let Some(result) = state.result.take() {
             Poll::Ready(result)
         } else {
+            // SAFETY:
+            // `callee`, being &[u8], is a readable sequence of bytes and therefore can be passed to ic0.call_new.
+            // `method`, being &str, is a readable sequence of bytes and therefore can be passed to ic0.call_new.
+            // `callback` is a function with signature (env : i32) -> () and therefore can be called as both reply and reject fn for ic0.call_new.
+            // `state_ptr` is a pointer created via Arc::into_raw, and can therefore be passed as the userdata for `callback`.
+            // `args`, being a &[u8], is a readable sequence of bytes and therefore can be passed to ic0.call_data_append.
+            // `cleanup` is a function with signature (env : i32) -> () and therefore can be called as a cleanup fn for ic0.call_on_cleanup.
+            // `state_ptr` is a pointer created via Arc::into_raw, and can therefore be passed as the userdata for `cleanup`.
+            // ic0.call_perform is always safe to call.
+            if state.waker.is_none() {
+                let callee = state.id.as_slice();
+                let method = &state.method;
+                let args = &state.arg;
+                let payment = state.payment;
+                let payment128 = state.payment128;
+                let err_code = unsafe {
+                    ic0::call_new(
+                        callee.as_ptr() as i32,
+                        callee.len() as i32,
+                        method.as_ptr() as i32,
+                        method.len() as i32,
+                        callback as usize as i32,
+                        state_ptr as i32,
+                        callback as usize as i32,
+                        state_ptr as i32,
+                    );
+
+                    ic0::call_data_append(args.as_ptr() as i32, args.len() as i32);
+                    {
+                        if payment > 0 {
+                            // SAFETY: ic0.call_cycles_add is always safe to call.
+                            // This is called as part of the call_new lifecycle, and so will not trap.
+                            ic0::call_cycles_add(payment as i64);
+                        }
+                    }
+                    add_payment(payment128);
+                    ic0::call_on_cleanup(cleanup as usize as i32, state_ptr as i32);
+                    ic0::call_perform()
+                };
+
+                // 0 is a special error code meaning call_simple call succeeded.
+                if err_code != 0 {
+                    state.result = Some(Err((
```

Comment:
This is resolving the future without reporting Poll::Ready. This would indefinitely hang the operation as it will receive no more wake-up messages.

Reply:
Indeed. I've fixed that and point out that I'm storing a (cloned) result into `state.result`, too, so that repeated calls to `poll` do not attempt to fire the call again, but rather return `Poll::Ready(result)`.

Reply:
Can't hurt, but for future reference, the result of calling poll again after returning Poll::Ready is allowed to do anything incoherent as long as it is not UB.